### ISSUE TITLE
Stackable tools

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#11127]
## Problem

Tools encorce a maximum stack size of 1. That's a limitation. It also causes makes tools slightly annoying to use in the code, since its an awkward special case. 

(Note: I define a “tool” to be any item registered with `minetest.register_tool`.)

## Suggestion

I suggest the following, written by @kahrl:

> It was I who originally wrote the ItemStack deserialization, but now I think that the engine (and builtin) should not enforce a stack_max of 1 for tools at all. Instead it should allow mods and subgames to have stackable tools if they want to simply by overriding stack_max.
>
> Stackable tools could work like in e.g. Junk Jack: if a tool with item count greater than 1 wears out, the count is decremented by one and wear is reset to 0, basically the worn out tool is replaced by a full-durability tool from the stack. So, a tool with item count C can be thought of as one currently active tool plus (C-1) full-durability backup tools.

> Of course there would have to be other considerations, such as how add_item would deal with such stacks.

https://github.com/minetest/minetest/issues/4497#issuecomment-244779354

I think that is great idea. It's simple, and it makes complete sense. I also think this could simplify a lot of tool-related code as well.

That's not a complete suggestion, and some questions still need to be answered (which I will do):

* What happens if in the inventory screen you rightclick a tool stack with a worn tool in it? You will pick up the worn tool. This means the leftover stack will NOT have wear
* What happens if you middle-click the same stack? You will pick up 10 tools, including the worn one
* What happens if you throw away a single tool from a tool? You will throw away the worn one

Long story short, for all player interactions, the worn tool should be considered to be the one that is actually used (like kahrl suggested), which implies that it also affects all inventory interaction, also functions like `remove_item`. I see it as the worn tool being on the “top” of an actual stack (as in “data structure”), and when you pick up items from it, you will always pick from the top first.

Finally, I like to stress that the *default* `stack_max` of tools should remain 1, and is only different if it was explicitly defined by the mod. Otherwise, a LOT of games and mods would break if tools suddenly default to a stack size of 99.

**Comment by voxel-minetest** [comment]
But if so, as an opt-in please, else it would be op...

**Comment by Wuzzy2** [comment]
Yes, obviously.

> Finally, I like to stress that the default stack_max of tools should remain 1, and is only different if it was explicitly defined by the mod.

In other words, this means that all existing games would remain unchanged. Games would need to **explicitly** change the `stack_max` to get stackable tools.

**Comment by Ghurir** [comment]
I this a thing
